rpk debug: Add 'bundle' command

0x5d opened the pull request:
'rpk debug bundle' collects environment data that can help debug and diagnose issues with a redpanda cluster, a broker, or the machine it's running on. It then bundles the collected data into a zip file.

The following are the data sources that are bundled in the compressed file:

 - Kafka data: Number of brokers, consumer groups, topics & partitions, topic
  configuration and replication factor.
 - /proc/cpuinfo: CPU information like make, core count, cache, frequency.
 - /proc/interrupts: IRQ distribution across CPU cores.
 - redpanda logs: The redpanda logs written to journald. If --logs-since or
  --logs-until are passed, then only the logs within the resulting time frame
  will be included.
 - redpanda configuration: The redpanda configuration file (redpanda.yaml).
 - Resource usage data: CPU usage percentage, free memory available for the
  redpanda process.
 - Clock drift: The ntp drift as reported by 'clockdiff'. Needs CAP_NET_RAW,
  so it's only included if 'rpk debug bundle' is run as root.
 - Network config: As reported by 'ifconfig'.
 - DNS: The DNS info as reported by 'dig', using the hosts in
  /etc/resolv.conf.
 - Virtual memory stats: As reported by 'vmstat'.
 - Socket info: The active sockets data output by 'ss'.
 - Running process info: As reported by 'top'.
 - Data directory structure: The 'tree' output describing the data directory
  contents.

All these "steps" are run concurrently and in a "fire and forget" manner: if any of them fails, the error will be written to 'errors.txt' within the ZIP bundle, but won't stop other steps from running.

twmb commented on src/go/rpk/pkg/system/syslog/syslog.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+const (
+	// Taken from `man syslog`
```
`man 2 syslog`

I think my man pages default to 3.

0x5d replied:
Fixed.

twmb commented on src/go/rpk/pkg/system/syslog/syslog.go:
```diff
@@ -0,0 +1,46 @@
+	// to by bufp.  The call reads the last len bytes from  the  log  buffer  (nondestruc‐
+	// tively),  but  will  not  read more than was written into the buffer since the last
+	// "clear ring buffer" command.  The call returns the number of bytes read.
+	syslog_action_read_all int = 3
```
I think since this is a c&p of the man docs, maybe the doc comment can be left out, and replaced with `See man 2 syslog`.  It makes more sense in context in the man page itself, since `len` is underlined and clearly refers to the arg to the klogctl function.

Same comment for size_buffer below.

I can't quite tell whether snake_case is good here... definitely violates Go naming guidelines, but it does match directly to the man page. I think if we're going per the man page it may make sense to have this be uppercased, otherwise maybe nix the underscore?

twmb replied:
Actually, these constants can be removed and replaced with the unix package equivalents:
unix.SYSLOG_ACTION_READ_ALL
unix.SYSLOG_ACTION_SIZE_BUFFER

0x5d replied:
Ah, check it out! Nice find, thanks.

0x5d replied:
Fixed.

twmb commented on src/go/rpk/pkg/system/syslog/syslog.go:
```diff
@@ -0,0 +1,46 @@
+
+import (
+	"fmt"
+	"syscall"
```
This should use the x/sys/unix package, as syscall is deprecated.

0x5d replied:
Fixed.

twmb commented on src/go/rpk/pkg/system/syslog/syslog.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+func ReadAll() ([]byte, error) {
+	bufSize, err := syscall.Klogctl(syslog_action_size_buffer, nil)
```
This can be https://pkg.go.dev/golang.org/x/sys@v0.0.0-20210817142637-7d9622a276b7/unix#Klogctl

0x5d replied:
Fixed.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,552 @@
+)
+
+// Use the same date specs as journalctl (see `man journalctl`).
+const timeHelpText = `Date specifications should be of the format "2012-10-30
```
This help text is way too long to be displayed alongside a flag, I think we can just say `(journalctl date format; e.g. YYYY-MM-DD)`. We may want to extend the since / until to more than just journalctl down the line, but I think this is fine to start since we expect debug to be used by us (so if we extend the format later ... we break ourselves).

0x5d replied:
Fixed.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,552 @@
+	filename := fmt.Sprintf("%d-bundle.zip", timestamp)
+	f, err := fs.OpenFile(
+		filename,
+		os.O_CREATE|os.O_WRONLY,
```
`os.O_EXCL` as well, for paranoia

0x5d replied:
Per the previous discussion with @senior7515, we will allow overwriting existing files, so O_EXCL isn't necessary anymore.

twmb replied:
I thought that was only for the bundle. But, there's not much risk to having two bundles in the same second.

0x5d replied:
I got my PRs mixed up hahahahaha. I just came back here after checking the comments in the ballast tuner PR 🤦🏽 . Yeah, I think this shouldn't be a problem here either.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,552 @@
+
+type stepParams struct {
+	fs            afero.Fs
+	m             *sync.Mutex
```
This can just be a `sync.Mutex`, no need for a pointer

0x5d replied:
Since this gets passed to several other functions, wouldn't that cause the lock to be copied?

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,552 @@
+
+	grp := multierror.Group{}
+
+	m := &sync.Mutex{}
```
This can be left out, especially if switching the field in `stepParams` from a pointer to a non-pointer

twmb replied:
This can be left out.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,552 @@
+
+	log.Infof("Debug bundle saved to '%s'", filename)
+
+	manifestFilename := fmt.Sprintf("%d-bundle-manifest.txt", timestamp)
```
Is the manifest important? I think it's easier for users to just send us one file rather than two, and the manifest we receive may not reflect reality -- I usually prefer not to have a separate file list. If we want something like this we may want to create a tarball, which allows for file inspection without decompressing everything.

0x5d replied:
@esteban especifically asked for this.

twmb replied:
I've pinged @esteban a bit and I think we need to define the manifest file. A simple list of files in the zip itself is not useful (because we can look at the list of files in a zip without decompressing it). We'd want the manifest to have more expanded information about the commands run,

"""
a collection of files and commands run by the bundle command with finer detail about the results of the execution of each command, e.g. did we get stderr? what was the exit code? if we run a command why it failed passing eth0 instead of en0? and so on
"""

This could be either a csv or json (probably json, since it's easier to add to).

We currently don't have a definition of a manifest file, and I don't think we want to try to spec out a manifest right now when we have so few bundles being generated and aren't running _that_ much. We can spec out the manifest file later, but I think we should drop the manifest from this PR itself. When we do have an expanded manifest later, the json type manifest can be bundled in the zip itself as well.

0x5d replied:
Let me check with him to see if we can drop it or if we need to define the manifest well and include it in this PR.

0x5d replied:
Manifest: removed.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,552 @@
+
+func saveConfig(ps *stepParams, conf *config.Config) step {
+	return func() error {
+		bs, err := yaml.Marshal(conf)
```
We need to strip any sasl password here, and potentially other things (I don't think hostnames are sensitive, but some people may not like it).

0x5d replied:
That's a good point, and an easy change. @esteban and I discussed the hostnames too, but agreed that we can add fine-grained redaction flags that adjust to each user's policies.

0x5d replied:
No redaction filters for now, but SASL credentials are now redacted.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,552 @@
+		return writeCommandOutputToZip(
+			ps,
+			"redpanda.log",
+			"journalctl",
```
Do we have concerns about bundling sensitive information here? I don't think redpanda dumps sensitive info, just checking.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,552 @@
+		return writeCommandOutputToZip(
+			ps,
+			"data-dir-tree.txt",
+			"tree",
```
`find` might be more ubiquitous. I don't have tree on my system.

0x5d replied:
Good tip, thanks!

senior7515 replied:
Wouldn’t this be more robust if we didn’t use shell for simple things like recursive directory traversal. Effectively a tree traversal

0x5d replied:
Removed and reimplemented by doing a dir walk in Go, that also follows symlinks that point to dirs.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,552 @@
+		10*time.Second,
+		"The maximum amount of time to wait for child commands (ss, ifconfig, "+
+			"dig, etc.) to be executed. "+
+			"The value passed is a sequence of decimal numbers, each with optional "+
```
I don't think this line nor the next two are necessary for the help text output (and it makes the help text too long)

0x5d replied:
I think at least the units part is important. Time can be specified in many formats. How about
```go
		"The maximum amount of time to wait for child commands (ss, ifconfig, "+
			"dig, etc.) to be executed. E.g.  '10000ms', '30s', '1m', '0.6h'",
```
?

twmb replied:
I think it could be very succinct:
```go
how long to wait for child commands to execute (e.g. '100ms', '30s', '1m')
```
your option is good too

0x5d replied:
Fixed.

twmb reviewed (commented):
Only important comments for now -- not trying to get the output perfect since this is super useful for us and we can improve this command as we go along.

senior7515 commented:
A few design things. 

1 what are the limits we are setting for logs. Last 100mb? 
2 for all shell commands are there go libraries we could use so they are more predictable in the light of things like kubernetes 
3 I saw no integration with the /admin endpoints for example is the node ready 
4 we should integrate the metrics that seastar reports for stalls, memory use, raft leadership count, writing to disk errors, etc. we expose them via the admin api

0x5d commented:
> A few design things.
> 
> 1 what are the limits we are setting for logs. Last 100mb?

Good point. Right now the default is "since the beginning". journalctl doesn't offer size-based filters, so how about a time-based default, like the last 3 days?

> 2 for all shell commands are there go libraries we could use so they are more predictable in the light of things like kubernetes

I will find them and replace them where I can.

> 3 I saw no integration with the /admin endpoints for example is the node ready
> 4 we should integrate the metrics that seastar reports for stalls, memory use, raft leadership count, writing to disk errors, etc. we expose them via the admin api

For 3. & 4.: There are still some things that will be added after this first version. The idea was to release something quickly and after that open the door for other people (the support team, other users, me) to contribute and improve it. The /admin info could be one of them.

0x5d commented:
Reverting to draft while the manifest questions are solved. I'll also be reimplementing some steps using plain go instead of _shelling out_.

senior7515 commented:
> > A few design things.
> > 1 what are the limits we are setting for logs. Last 100mb?
> 
> Good point. Right now the default is "since the beginning". journalctl doesn't offer size-based filters, so how about a time-based default, like the last 3 days?


It needs size based as we are exporting these. You can stream read in reverse until beginning of time or some size based like 250MB before compression 


> 
> > 2 for all shell commands are there go libraries we could use so they are more predictable in the light of things like kubernetes
> 
> I will find them and replace them where I can.
> 

great


> > 3 I saw no integration with the /admin endpoints for example is the node ready
> > 4 we should integrate the metrics that seastar reports for stalls, memory use, raft leadership count, writing to disk errors, etc. we expose them via the admin api
> 
> For 3. & 4.: There are still some things that will be added after this first version. The idea was to release something quickly and after that open the door for other people (the support team, other users, me) to contribute and improve it. The /admin info could be one of them.


Makes sense. Though dumping them on a flat text file seems critical and easy, part of MVP? For example cpu usage and other while important less critical than a text dump snapshot of /metrics

esteban commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,549 @@
+  redpanda process.
+ - Clock drift: The ntp drift as reported by 'clockdiff'. Needs CAP_NET_RAW,
+  so it's only included if 'rpk debug bundle' is run as root.
+ - Network config: As reported by 'ifconfig'.
```
any preference for ifconfig vs ip?  ifconfig has also been deprecated so perhaps might be good just to go with ip only.

0x5d replied:
I personally use `ip` but during our conversation you mentioned `if`, so I used that one instead. I might have been a typo, then. However, I'll be trying to replace all the cmds with Go code/ existing libs, so I don't think I'll use either.

0x5d replied:
Fixed.

0x5d commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,549 @@
+		return writeCommandOutputToZip(
+			ps,
+			"top.txt",
+			"top", "-n", "1",
```
How many iterations would be ok? Would 10 be fine?

0x5d replied:
Done

0x5d commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,549 @@
+	if err != nil {
+		return err
+	}
+	cmd.Stdout = wr
```
Done

0x5d commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,549 @@
+		return writeCommandOutputToZip(
+			ps,
+			"vmstat.txt",
+			"vmstat",
```
Done

0x5d commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,549 @@
+		return writeCommandOutputToZip(
+			ps,
+			"data-dir-tree.txt",
+			"find",
```
I tried reproducing this behavior in Go itself (following links). The output is a JSON map where the key is the file name and the values are objects holding the file info (size, owner, group, mode, etc.)

0x5d commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,549 @@
+	}
+}
+
+func saveCPUInfo(ps *stepParams) step {
+	return func() error {
+		bs, err := afero.ReadFile(ps.fs, "/proc/cpuinfo")
+		if err != nil {
+			return err
+		}
+		return writeFileToZip(ps, "proc/cpuinfo", bs)
+	}
+}
```
Done.

0x5d commented:
@twmb @esteban @senior7515 I believe I have addressed all the comments & made the requested changes. Since the last time you reviewed:
- I added the lspci, dmidecode, ss output to the bundle.
- Reimplemented some "steps" (data dir structure & ntp drift) in Go.
- Improved the command's help text.
- Added the GET /metrics output to the bundle.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+	ps.m.Lock()
+	defer ps.m.Unlock()
+
+	ctx, _ := context.WithTimeout(context.Background(), ps.timeout)
```
Always worth keeping the `cancel` so as to `defer cancel()`, it's not too critical here, but it frees up timers in Go's runtime timer heap.

0x5d replied:
Fixed.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+
+	err = cmd.Wait()
+	if err != nil {
+		if strings.Contains(err.Error(), "broken pipe") {
```
If was looking at the stdlib and it looks like broken pipe is returned if the fd's returned from os.Pipe fail, but looking at the os.Exec code, it reads as if the error it saves is the first one encountered -- which for us should be the write returned from the `limitedWriter`. Were you able to hit this broken pipe case?

0x5d replied:
The 2nd one - when the output exceeded the limit and limitedWriter errored.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+
+	l.accumBytes += n
+
+	if l.accumBytes >= l.limitBytes {
```
If we want to stop at the limit, we should check `l.accumBytes + len(p) >= l.limitBytes` before the write (and truncate p accordingly, or not perform the write).

0x5d replied:
Sounds good. I intentionally left it like this (not strict) because each write is a block long, which on many machines is 4kB. But I agree, let's make it strict.

0x5d replied:
Fixed.

twmb replied:
I think this will fail if the limit is 100 and we write exactly 100 bytes, instead we may want to change the initial conditional to:
```go
if l.accumBytes+len(p) > l.limitBytes {
  p = ...
  limitReached = true
}
```
and then return err here `if limitReached`

0x5d replied:
Done.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+
+ - Data directory structure: A file describing the data directory's contents.
+
+ - redpanda configuration: The redpanda configuration file (redpanda.yaml).
```
`redpanda.yaml; sasl is stripped)`

0x5d replied:
Done.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+
+type stepParams struct {
+	fs      afero.Fs
+	m       *sync.Mutex
```
This can be `m sync.Mutex`.

0x5d replied:
Since this gets copied over to each "`step`", wouldn't that cause the mutex instance to be copied too, resetting it for each?

twmb replied:
`stepParams` itself is initialized as a pointer, so the mutex itself is never copied. `go vet` would warn against mutex copies, which can help ensure this works (https://play.golang.org/).

0x5d replied:
Done

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+	Id   string `json:"id"`
+}
+
+func saveKafkaMetadata(ps *stepParams, admin sarama.ClusterAdmin) step {
```
I think this is fine for now but it might be simpler to issue rpk commands directly and just capture the output (rpk cluster offsets, etc), and it'd give us incentive to improve that output.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+		Group    string    `json:"group"`
+	}
+	var walk func(string, map[string]*fileInfo) error
+	walk = func(root string, files map[string]*fileInfo) error {
```
I think this `walk` function should be defined outside of `saveDataDirStructure`, potentially just have it be `walkRedpandaDataDir`. That'd eliminate the need for the `var walk ...` just above: closures cannot refer to themselves, but top level functions can. It'd also help eliminate some nesting.

0x5d replied:
Sounds good. This function started small and then grew a lot haha.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+		Mode     string    `json:"mode"`
+		Symlink  string    `json:"symlink,omitempty"`
+		Error    string    `json:"error,omitempty"`
+		Modified time.Time `json:"modified"`
```
json isn't great at encoding a time.Time; we may want to `.Format` this.

0x5d replied:
Done

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+					return nil
+				}
+
+				files[path] = &fileInfo{}
```
How about
```go
i := new(fileInfo)
files[path] = i
```
and then using `i` below rather than `files[path]`?

0x5d replied:
Done

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+				// The user and group are only available through the
+				// underlying syscall object.
+				sys, ok := info.Sys().(*syscall.Stat_t)
+				if ok && sys != nil {
```
Is there ever a case where this returns `ok` but `sys` is nil?

0x5d replied:
Fixed

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+				// underlying syscall object.
+				sys, ok := info.Sys().(*syscall.Stat_t)
+				if ok && sys != nil {
+					u, err := user.LookupId(fmt.Sprint(sys.Uid))
```
I think `strconv.Itoa` is more common for this type of thing.

0x5d replied:
Uid is a uint32, so I went for `fmt.Sprint` instead.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+				if ok && sys != nil {
+					u, err := user.LookupId(fmt.Sprint(sys.Uid))
+					if err == nil {
+						files[path].User = u.Name
```
If ere is nil, maybe `User = fmt.Sprintf("user lookup %d failed: %v", sys.Uid, err)`, and similar thing for the group?

0x5d replied:
Done

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+				// If the file it points to is a directory, follow it and then
+				// call `walk` using it as the root.
+				isSymlink := info.Mode().Type()&fs.ModeSymlink != 0
+				if isSymlink {
```
`if !isSymlink { return nil }`
and then un-indent the block of logic that follows?

0x5d replied:
Done

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+	walk = func(root string, files map[string]*fileInfo) error {
+		return filepath.WalkDir(
+			root,
+			func(path string, d fs.DirEntry, readErr error) error {
```
I think we may want to never return `err` from this function. Returning an error when we fail below stops walking entirely, _but_ if we do not return an error (and instead save errors, like you do currently), we can allow `WalkDir` to continue and grab information on a best effort basis.

0x5d replied:
Hmm. The signature is mandatory (`fs.WalkDirFunc`), but it doesn't really return an error other than `fs.SkipDir` when we know we can just skip the dir since it can't be read.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+		)
+	}
+	return func() error {
+		files := map[string]*fileInfo{}
```
`files := make(...)` is a bit more idiomatic. One of our more stringent linters lints against this.

0x5d replied:
Done.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+// Queries 'pool.ntp.org' and writes a file with the reported RTT, time & precision.
+func saveNTPDrift(ps *stepParams) step {
+	return func() error {
+		host := "pool.ntp.org"
```
`const host =`

0x5d replied:
Done.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+			ps,
+			"redpanda.log",
+			logsLimitBytes,
+			"journalctl",
```
Can we make a followup issue for redpanda not to dump sasl configuration on startup? (maybe it doesn't already, but it's worth triple checking, because I remember redpanda straight up dumping the redpanda.yaml file to logs)

0x5d replied:
For sure, feel free to create it!

twmb replied:
https://github.com/vectorizedio/redpanda/issues/2247

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,842 @@
+		return writeCommandOutputToZip(
+			ps,
+			"top.txt",
+			"top", "-n", "10", "-H",
```
`-n 100` may be a bit better. If redpanda is stuck, it may not be using much CPU, but maybe it has like 100G of memory in use which would be significant for us to know.

0x5d replied:
-n is the number of iterations, not lines 🙏🏽

dotnwat commented:
You can also grab `redpanda configuration: The redpanda configuration file (redpanda.yaml).` from the running redpanda process for even more precision. I think that once this tool is in place it will be very useful to provide an admin http endpoint that can be scraped locally for all sorts of interesting stuff. Might be useful to have a `--core` option to grab a core dump with `gcore` and compress it and upload some where. Ok maybe that is v2.

0x5d commented:
> You can also grab `redpanda configuration: The redpanda configuration file (redpanda.yaml).` from the running redpanda process for even more precision. 

Oh, cool, I didn't know that was there already. `9644/v1/config`, right?

> I think that once this tool is in place it will be very useful to provide an admin http endpoint that can be scraped locally for all sorts of interesting stuff. Might be useful to have a `--core` option to grab a core dump with `gcore` and compress it and upload some where. Ok maybe that is v2.

I agree.

senior7515 commented:
@twmb  + @0x5d  - what can we do to get this merged. this is really valuable as is, but needs to get through code review. I think we can postpone other features to v2?

senior7515 commented:
@bmansheim - let's aim to have docs for this for the next release too. You can run it locally once it releases and as v1, list out the things it collects as a table

0x5d commented:
> You can also grab `redpanda configuration: The redpanda configuration file (redpanda.yaml).` from the running redpanda process for even more precision.

@dotnwat how can I do that? Is it `9644/config?`

I think that once this tool is in place it will be very useful to provide an admin http endpoint that can be scraped locally for all sorts of interesting stuff. Might be useful to have a `--core` option to grab a core dump with `gcore` and compress it and upload some where. Ok maybe that is v2.

Yeah. I think the structure of the code lends itself to adding new files to the bundle fairly easily.



> @twmb + @0x5d - what can we do to get this merged. this is really valuable as is, but needs to get through code review. I think we can postpone other features to v2?

I agree. Although I can finish going over @twmb 's suggestions tomorrow, since they're very specific.

senior7515 commented:
> I agree. Although I can finish going over @twmb 's suggestions tomorrow, since they're very specific.


sounds good

0x5d commented on src/go/rpk/pkg/api/admin/api_metrics.go:
```diff
@@ -0,0 +1,18 @@
+import "net/http"
+
+func (a *AdminAPI) PrometheusMetrics() ([]byte, error) {
+	res := []byte{}
```
Fixed.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle.go:
```diff
@@ -0,0 +1,853 @@
+
+			info, err := d.Info()
+			if err != nil {
+				files[path].Error = err.Error()
```
this could be `i.Error =`

0x5d replied:
Fixed.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle_test.go:
```diff
@@ -0,0 +1,71 @@
+			for remaining > 0 {
+				mult := math.Min(remaining, 1)
+				bs := make([]byte, int(block*mult))
+				_, err := rand.Read(bs)
+				require.NoError(st, err)
```
No need for these two lines, we aren't testing we can read, and we aren't checking the random contents.

0x5d replied:
I like to test with random data when possible (you never know what weird corner case you'll find), but this is simple enough.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle_test.go:
```diff
@@ -0,0 +1,71 @@
+				require.NoError(st, err)
+
+				var n int
+				n, writeErr = lim.Write(bs)
```
I think you want to break if this write fails, because `writeErr` could fail the first time, and then this loop.

0x5d replied:
Good point, done.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle_test.go:
```diff
@@ -0,0 +1,71 @@
+	tests := []struct {
+		name          string
+		limit         int
+		blocksToWrite float64
```
I think we could avoid floats and any rounding error they have if the two tests below are:

1) limit 3, blocksToWrite 2
2) limit 1, blocksToWrite 2

which would also remove the need for math.Min

0x5d replied:
I wanted to simulate non-aligned contents, but I think this is fine.

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle_test.go:
```diff
@@ -0,0 +1,71 @@
+				var n int
+				n, writeErr = lim.Write(bs)
+				written += n
+				remaining -= 1
```
--

twmb commented on src/go/rpk/pkg/cli/cmd/debug/bundle_test.go:
```diff
@@ -0,0 +1,71 @@
+			expected := tt.limit
+			if int(tt.blocksToWrite*block) > tt.limit {
+				require.EqualError(st, writeErr, "output size limit reached")
+			} else if int(tt.blocksToWrite*block) < tt.limit {
```
<=

0x5d replied:
If `if int(tt.blocksToWrite*block) == tt.limit`, then `expected := tt.limit`, which is the init value.

0x5d replied:
I changed it nonetheless.

0x5d commented on src/go/rpk/pkg/cli/cmd/debug/bundle_test.go:
```diff
@@ -0,0 +1,71 @@
+		t.Run(tt.name, func(st *testing.T) {
+			writer := ioutil.Discard
+			lim := &limitedWriter{
+				w:          writer,
```
Good catch.

twmb reviewed (dismissed):
lgtm!

dotnwat commented:
> > You can also grab `redpanda configuration: The redpanda configuration file (redpanda.yaml).` from the running redpanda process for even more precision.
> 
> @dotnwat how can I do that? Is it `9644/config?`
> 
> I think that once this tool is in place it will be very useful to provide an admin http endpoint that can be scraped locally for all sorts of interesting stuff. Might be useful to have a `--core` option to grab a core dump with `gcore` and compress it and upload some where. Ok maybe that is v2.
> 
> Yeah. I think the structure of the code lends itself to adding new files to the bundle fairly easily.
> 
> > @twmb + @0x5d - what can we do to get this merged. this is really valuable as is, but needs to get through code review. I think we can postpone other features to v2?
> 
> I agree. Although I can finish going over @twmb 's suggestions tomorrow, since they're very specific.

Yep, definitely cool to wait until v2!

mattschumpert commented:
Related: https://github.com/redpanda-data/redpanda/issues/4911
